source/common/network/address_impl.cc

```diff
@@ -88,27 +69,48 @@ Address::InstanceConstSharedPtr addressFromSockAddr(const sockaddr_storage& ss,
     RELEASE_ASSERT(ss_len == 0 || static_cast<unsigned int>(ss_len) >=
                                       offsetof(struct sockaddr_un, sun_path) + 1,
                    "");
-    return std::make_shared<Address::PipeInstance>(sun, ss_len);
+    return Address::InstanceFactory::createInstancePtr<Address::PipeInstance>(sun, ss_len);
   }
   default:
-    throw EnvoyException(fmt::format("Unexpected sockaddr family: {}", ss.ss_family));
+    return absl::InvalidArgumentError(fmt::format("Unexpected sockaddr family: {}", ss.ss_family));
   }
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
+Address::InstanceConstSharedPtr addressFromSockAddrOrThrow(const sockaddr_storage& ss,
+                                                           socklen_t ss_len, bool v6only) {
+  StatusOr<InstanceConstSharedPtr> address = addressFromSockAddr(ss, ss_len, v6only);
+  if (!address.ok()) {
+    throw EnvoyException(address.status().ToString());
+  }
+  return *address;
+}
+
+Address::InstanceConstSharedPtr
+getAddressFromSockAddrOrDie(const sockaddr_storage& ss, socklen_t ss_len, os_fd_t fd, bool v6only) {
+  // Set v6only to false so that mapped-v6 address can be normalize to v4
+  // address. Though dual stack may be disabled, it's still okay to assume the
+  // address is from a dual stack socket. This is because mapped-v6 address
+  // must come from a dual stack socket. An actual v6 address can come from
+  // both dual stack socket and v6 only socket. If |peer_addr| is an actual v6
+  // address and the socket is actually v6 only, the returned address will be
+  // regarded as a v6 address from dual stack socket. However, this address is not going to be
+  // used to create socket. Wrong knowledge of dual stack support won't hurt.
+  StatusOr<Address::InstanceConstSharedPtr> error_or_addr =
+      Address::addressFromSockAddr(ss, ss_len, v6only);
+  if (!error_or_addr.ok()) {
+    PANIC(fmt::format("Invalid address for fd: {}", fd));
```

Comment:
strcat in `error_or_addr.message()`

also per earlier comment, I think you can just call this local variable `address`

Reply:
Thanks, will clean up.